### PR TITLE
Speed up queries by optimizing Genes, Transcripts and Exons table indexes

### LINKED ISSUE #103: Speed up transcripts and exons selects
Especially when it comes to evaluate the coverage

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
-  Fix #103 

### How to test:
- Launch a sample coverage query on transcripts and exons

### Expected outcome:
- [x] The query should be quick(er)

### Review:
- [x] Code approved by HS
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by northwestwitch** [comment]
- This brach, transcripts query is fast enough:

image

- Exons query is still slow, but feels faster than before:

- image

I guess it can be improved, but the time takes into account also the computation of coverage and completeness over all the exons. Well, this is the first improvement!

**Comment by northwestwitch** [comment]
Meh I'm not satisfied with the performance of the queries. I think one of the reasons why they're so slow is because of the usage of IN condition. Example:

```
2023-05-11 14:34:53,637 INFO sqlalchemy.engine.Engine SELECT exons.id AS exons_id, exons.chromosome AS exons_chromosome, exons.start AS exons_start, exons.stop AS exons_stop, exons.ensembl_id AS exons_ensembl_id, exons.ensembl_gene_id AS exons_ensembl_gene_id, exons.build AS exons_build
FROM exons
WHERE exons.ensembl_gene_id IN (%s) AND exons.build = %s
```

I'll try to replace the INs with table joins instead!

**Comment by northwestwitch** [comment]
Ok now I am satisfied with tables and indexes. The time to retrieve all exons from a gene lis quite fast:

image

While the time to compute coverage and completeness over the exons of the same gene is still quite long (10 seconds for a gene with a lot of exons):

image

But I'll try to address this in another PR!

File: src/chanjo2/models/sql_models.py [hunk 1]
```diff
@@ -68,16 +74,23 @@ class Transcript(Base):
...
+    refseq_ncrna = Column(String(24), nullable=True)
     refseq_mane_select = Column(String(24), nullable=True, index=True)
     refseq_mane_plus_clinical = Column(String(24), nullable=True, index=True)
-    ensembl_gene_id = Column(String(24), nullable=False, index=True)
+    ensembl_gene_id = Column(String(24), ForeignKey("genes.ensembl_id"), nullable=False)
     build = Column(
         Enum(Builds, values_callable=lambda x: Builds.get_enum_values()), index=True
     )
 
+    genes = relationship(
```
Comment (northwestwitch) [inline src/chanjo2/models/sql_models.py:87]: This is speeding up the whole processing of retrieving transcripts and exons by allowing to perform queries on joined tables transcripts and genes or exons and genes

File: src/chanjo2/models/sql_models.py [hunk 2]
```diff
@@ -88,7 +101,14 @@ class Exon(Base):
     start = Column(Integer, nullable=False)
     stop = Column(Integer, nullable=False)
     ensembl_id = Column(String(24), nullable=False, index=False)
-    ensembl_gene_id = Column(String(24), nullable=False, index=True)
+    ensembl_gene_id = Column(String(24), ForeignKey("genes.ensembl_id"), nullable=False)
     build = Column(
         Enum(Builds, values_callable=lambda x: Builds.get_enum_values()), index=True
     )
+
+    genes = relationship(
```
Comment (northwestwitch) [inline src/chanjo2/models/sql_models.py:109]: Same this one

File: src/chanjo2/crud/intervals.py [hunk 3]
```diff
@@ -108,42 +72,22 @@ def get_genes(
 ) -> List[SQLGene]:
     """Return genes according to specified fields."""
     genes: query.Query = db.query(SQLGene)
+
     if ensembl_ids:
-        genes: query.Query = _filter_intervals_by_ensembl_ids(
-            intervals=genes, interval_type=SQLGene, ensembl_ids=ensembl_ids
-        )
+        genes = genes.filter(SQLGene.ensembl_id.in_(ensembl_ids))
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:77]:
```suggestion
        genes: query.Query = genes.filter(SQLGene.ensembl_id.in_(ensembl_ids))
```

File: src/chanjo2/crud/intervals.py [hunk 4]
```diff
@@ -108,42 +72,22 @@ def get_genes(
...
     elif hgnc_ids:
-        genes: query.Query = _filter_intervals_by_hgnc_ids(
-            intervals=genes, interval_type=SQLGene, hgnc_ids=hgnc_ids
-        )
+        genes = genes.filter(SQLGene.hgnc_id.in_(hgnc_ids))
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:79]:
```suggestion
        genes: query.Query = genes.filter(SQLGene.hgnc_id.in_(hgnc_ids))
```

File: src/chanjo2/crud/intervals.py [hunk 5]
```diff
@@ -108,42 +72,22 @@ def get_genes(
...
     elif hgnc_symbols:
-        genes: query.Query = _filter_intervals_by_hgnc_symbols(
-            intervals=genes, interval_type=SQLGene, hgnc_symbols=hgnc_symbols
-        )
-    genes: query.Query = _filter_intervals_by_build(
+        genes = genes.filter(SQLGene.hgnc_symbol.in_(hgnc_symbols))
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:81]:
```suggestion
        genes: query.Query = genes.filter(SQLGene.hgnc_symbol.in_(hgnc_symbols))
```

File: src/chanjo2/crud/intervals.py [hunk 6]
```diff
@@ -108,42 +72,22 @@ def get_genes(
...
+
+    genes = _filter_intervals_by_build(
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:83]:
```suggestion
    genes: query.Query = _filter_intervals_by_build(
```

File: src/chanjo2/crud/intervals.py [hunk 7]
```diff
@@ -179,36 +123,22 @@ def get_gene_intervals(
...
-        ensembl_gene_ids: List[str] = [gene.ensembl_id for gene in genes]
+) -> List[Union[SQLTranscript, SQLExon]]:
+    """Retrieve transcripts or exons from a list of genes."""
 
+    intervals: query.Query = db.query(interval_type).join(SQLGene)
     if ensembl_ids:
-        intervals: query.Query = _filter_intervals_by_ensembl_ids(
-            intervals=intervals, interval_type=interval_type, ensembl_ids=ensembl_ids
-        )
+        intervals = intervals.filter(interval_type.ensembl_id.in_(ensembl_ids))
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:131]:
```suggestion
        intervals: query.Query = intervals.filter(interval_type.ensembl_id.in_(ensembl_ids))
```

File: src/chanjo2/crud/intervals.py [hunk 8]
```diff
@@ -179,36 +123,22 @@ def get_gene_intervals(
...
     elif ensembl_gene_ids:
-        intervals: query.Query = _filter_intervals_by_ensembl_gene_ids(
-            intervals=intervals,
-            interval_type=interval_type,
-            ensembl_gene_ids=ensembl_gene_ids,
+        intervals = intervals.filter(
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:133]:
```suggestion
        intervals: : query.Query = intervals.filter(
```

File: src/chanjo2/crud/intervals.py [hunk 9]
```diff
@@ -179,36 +123,22 @@ def get_gene_intervals(
...
+            interval_type.ensembl_gene_id.in_(ensembl_gene_ids)
         )
+    elif hgnc_ids:
+        intervals = intervals.filter(SQLGene.hgnc_id.in_(hgnc_ids))
+    elif hgnc_symbols:
+        intervals = intervals.filter(SQLGene.hgnc_symbol.in_(hgnc_symbols))
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/intervals.py:139]:
```suggestion
        intervals: query.Query = intervals.filter(SQLGene.hgnc_symbol.in_(hgnc_symbols))
```

File: src/chanjo2/models/sql_models.py [hunk 10]
```diff
@@ -68,16 +74,23 @@ class Transcript(Base):
...
     build = Column(
         Enum(Builds, values_callable=lambda x: Builds.get_enum_values()), index=True
     )
 
+    genes = relationship(
+        "Gene",
+        primaryjoin="Transcript.ensembl_gene_id==Gene.ensembl_id",
+    )
+
+    __table_args__ = (Index("txidx_ensembl_gene_build", "ensembl_gene_id", "build"),)
```
Comment (henrikstranneheim) [inline src/chanjo2/models/sql_models.py:92]:
```suggestion
    __table_args__ = (Index("transcript_idx_ensembl_gene_build", "ensembl_gene_id", "build"),)
```

File: tests/src/chanjo2/endpoints/test_intervals.py [hunk 11]
```diff
@@ -214,6 +206,28 @@ def test_transcripts_multiple_filters(
...
+    """Test the endpoint that returns database transcripts without any filter."""
+
+    # GIVEN a populated demo database
+    # WHEN sending a request to the "transcripts" endpoint
+    data = {
+        "build": build,
+    }
+    response: Response = demo_client.post(endpoints.TRANSCRIPTS, json=data)
+
+    # THEN transcript documents should be returned
```
Comment (henrikstranneheim) [inline tests/src/chanjo2/endpoints/test_intervals.py:225]:
```suggestion
    # THEN transcript objects should be returned
```
Not documents right? If correct change for the other tests

**Review by henrikstranneheim — APPROVED** [review]
Awesome speedup 🥇 ! Well done

File: src/chanjo2/crud/intervals.py [hunk 12]
```diff
@@ -179,36 +123,22 @@ def get_gene_intervals(
...
 
-    intervals: query.Query = _filter_intervals_by_build(
-        intervals=intervals, interval_type=interval_type, build=build
-    )
+    intervals = intervals.filter(interval_type.build == build)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply src/chanjo2/crud/intervals.py:141]: You really are a fan of data types eh? 😄

File: tests/src/chanjo2/endpoints/test_intervals.py (hunk 11 above)
Comment (northwestwitch) [reply tests/src/chanjo2/endpoints/test_intervals.py:225]: Yes, objects. Will do, thanks!

File: src/chanjo2/crud/intervals.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply src/chanjo2/crud/intervals.py]: oh, you noticed 😄 ! It helps a lot when you are reviewing in so many repos and do not have the underlying code fresh in memory.

**Pull request merged**